### ISSUE TITLE
param grid search and custom parameters generation

### ISSUE THREAD

**Issue by doutriaux1** [LLNL/maestrowf#225]
I'm trying to do a parameter grid search. I'm trying to use a custom param generator

I run into a few issues:

1. Global.parameters expects all variables to have the same number of values

I went around this by putting the values in the env section (not ideal)

2. But I still cannot figure out how to access the label for each variable.

3. Finally even after successfully returning many parameters I only get ONE study run… I tried commenting out global.paramters but to no success.

I'm pasting my simple maestro ymal file and custom generator to help you reproduce.

```yaml
description:
    name: param grid sample test
    description: A sample parameter grid search study
env:
  variables:
    SIZE: [1,2,3,4]
    TRIAL: [4,5]
    ITERATION: [4,3,2]

study:
    - name: test_gen
      description: Build the serial version of LULESH.
      run:
          cmd: |
            echo $(TRIAL)_$(SIZE)_$(ITERATION)
          depends: []
#global.parameters:
#  SIZE:
#    values: [1,]
#    label: SIZE.%%
#  TRIAL:
#    values: [5,]
#    label: TRIAL.%%
#  ITERATION:
#    values: [5,]
#    label: ITERATION.%%
```

```python
"""An example file that produces a custom parameters for the LULESH example."""

from maestrowf.datastructures.core import ParameterGenerator
from sklearn.model_selection import ParameterGrid

def get_custom_generator(env, **kwargs):
    """
    Create a custom populated ParameterGenerator.

    This function recreates the exact same parameter set as the sample LULESH
    specifications. The point of this file is to present an example of how to
    generate custom parameters.

    :returns: A ParameterGenerator populated with parameters.
    """
    p_gen = ParameterGenerator()
    params = env.to_dict()["substitutions"]
    for key in ["OUTPUT_PATH", "SPECROOT"]:
        del(params[key])
    print(params)
    p = {}
    for k, val in params.items():
        if isinstance(val.value, (list,tuple)):
            p[k] = val.value
        else:
            p[k] = [val.value,]

    grid = ParameterGrid(p)
    p = {}
    for g in grid:
        for k in g:
            if not k in p:
                p[k] = [g[k],]
            else:
                p[k].append(g[k])
    print("OUT:", p)
    for k, val in p.items():
        p_gen.add_parameter(k, val, f"{k.upper()}.%%")

    return p_gen
```

**Comment by jsemler** [comment]
Hi @doutriaux1! Thanks for posting the ticket.

Maestro will only iterate based on the parameters. Since no parameters are defined Maestro will only run the step once.

Maestro requires that the lengths of all parameters be the same, so to run your variations you'll want to expand out your lists with all the combinations you want to run. The parameter block will look something like:

```
global.parameters:
    SIZE:
        values: [1, 2, 3, 4, 1, 2, 3, 4, ...]
        label: SIZE.%%
    TRIAL:
        values: [4, 4, 4, 4, 5, 5, 5, 5, ...]
        label: TRIAL.%%
    ITERATION:
       values: [4, 4, 4, 4, 4, 4, 4, 4, ...]
       label: ITERATION.%%
```

The above can be expanded for whichever combinations you want to run. If you are running
a lot of combinations ``pgen`` can be used to generate this list. Python's [itertools](https://docs.python.org/3/library/itertools.html) is a helpful package for expanding lists of variables easily and should help simplify your ``pgen`` setup.

**Comment by doutriaux1** [comment]
@jsemler but that's what I'm trying to use pgen with my custom generator. I'm not sure why it's not picking it up. It does print the values I expect.

**Comment by doutriaux1** [comment]
@jsemler the way i'm runnnig maestro is as follow:

```bash
maestro -p custom_generator.py run my_maestro.yaml
```

**Comment by jsemler** [comment]
@doutriaux1, try specifying ``pgen`` after the run argument.

```
maestro run my_maestro.yaml -p custom_generator.py
```

**Comment by doutriaux1** [comment]
@jsemler same result. It does see it though:

```bash
2020-02-12 10:47:39,274 - maestrowf.maestro:load_parameter_generator:105 - INFO - Loading custom parameter generator from '/home/doutriaux1/custom_my_maestro_custom_gen.py'
{'SIZE': <maestrowf.datastructures.environment.variable.Variable object at 0x7f1102cb90f0>, 'TRIAL': <maestrowf.datastructures.environment.variable.Variable object at 0x7f1102cb9048>, 'ITERATION': <maestrowf.datastructures.environment.variable.Variable object at 0x7f1102cb9198>}
OUT: {'ITERATION': [4, 4, 4, 4, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2], 'SIZE': [1, 1, 2, 2, 3, 3, 4, 4, 1, 1, 2, 2, 3, 3, 4, 4, 1, 1, 2, 2, 3, 3, 4, 4], 'TRIAL': [4, 5, 4, 5, 4, 5, 4, 5, 4, 5, 4, 5, 4, 5, 4, 5, 4, 5, 4, 5, 4, 5, 4, 5]}
2020-02-12 10:47:42,878 - maestrowf.datastructures.core.study:__init__:198 - INFO - OUTPUT_PATH = /home/doutriaux1/param_grid_sample_test_20200212-104739
2020-02-12 10:47:42,879 - maestrowf.datastructures.core.study:add_step:339 - INFO - Adding step 'test_gen' to study 'param grid sample test'...
2020-02-12 10:47:42,880 - maestrowf.datastructures.core.study:setup_workspace:378 - INFO - Setting up study workspace in '/home/doutriaux1/param_grid_sample_test_20200212-104739'
2020-02-12 10:47:42,881 - maestrowf.datastructures.core.study:setup_environment:388 - INFO - Environment is setting up.
2020-02-12 10:47:42,882 - maestrowf.datastructures.core.studyenvironment:acquire_environment:191 - INFO - Acquiring dependencies
2020-02-12 10:47:42,883 - maestrowf.datastructures.core.study:configure_study:426 - INFO -
------------------------------------------
Output path =               /home/doutriaux1/param_grid_sample_test_20200212-104739
Submission attempts =       1
Submission restart limit =  1
Submission throttle limit = 0
Use temporary directory =   False
Hash workspaces =           False
```

**Comment by FrankD412** [comment]
@doutriaux1 -- Thanks for the ticket.

That's interesting that pgen isn't being picked up. I'm currently working on looking into this -- I was able to reproduce this issue.

**Comment by jsemler** [comment]
When I try running the study, Maestro generates a script for the step with the following contents:

```
#!/bin/bash

echo [4, 5]_[1, 2, 3, 4]_[4, 3, 2]
```

It appears Maestro is replacing the variables before the parameters.

When I comment out the variable section and run with Maestro's example [LULESH pgen](https://github.com/LLNL/maestrowf/blob/develop/samples/parameterization/lulesh_custom_gen.py), the study runs with all the parameters.

**Comment by FrankD412** [comment]
Jessica is correct. The syntax for substituting variables is exactly the same as parameters. Maestro follows a progression as follows:

1. Substitute labels (combinations of variables, parameters, and literals)
2. Substitute variables (static, non-changing, user-specified key values)
3. Substitute the parameters for as many valid combinations as there are.

You're pgen script is doing the right thing, but Maestro never detects that the `test_gen` step has parameters because it substituted variables. Because the `pgen` file is Python code, you could have it load a file which contains the data or use some other method to obtain the data. You can even dictate where the data is loaded from in two ways:

1. Specify a `pargs` command line argument such that your custom pgen file knows the variable to expect. You would call maestro in the following way `maestro run -p pgen.py --pargs "DATA_PATH:<path to file>" test.yaml`. That allows you to rapidly change out the data file to run the same workflow on different data files and maintain the output separately.
2. Specify a variable in your specification 
```
env:
    DATA_PATH: <path to data file>
```
This method allows you to put the data file in a fixed place if it doesn't change all that often.

Hopefully this comment helps clarify some of the confusion.

**Comment by doutriaux1** [comment]
@FrankD412 are you saying generate a file that test_gen would pick up? What's I'm trying to achieve is to call test_gen numerous times. Otherwise I don't need maestro.

**Comment by doutriaux1** [comment]
@FrankD412 thanks for the explanations this afternoon. Following up here are the existing issues

1. when getting to the custom_gen function the yaml file is not yet copied in OUTPUT_PATH
2. fortunately the sys.argv still contains the path to original yaml file so i get it from there
3. unfortunately having global.parameters seems to NOT being ignored

my yaml file:

```yaml
description:
    name: param grid sample test
    description: A sample parameter grid search study

study:
    - name: test_gen
      description: Build the serial version of LULESH.
      run:
          cmd: |
            echo $(TRIAL)_$(SIZE)_$(ITERATION)
          depends: []
global.parameters:
  SIZE:
    values: [1,2,3]
    label: SIZE.%%
  TRIAL:
    values: [5,4,4]
    label: TRIAL.%%
  ITERATION:
    values: [5,5,5]
    label: ITERATION.%%
```

my python generator

```python
import sys
import os
from maestrowf.datastructures.core import ParameterGenerator
from sklearn.model_selection import ParameterGrid
import yaml
try:
    from yaml import CLoader as Loader, CDumper as Dumper
except ImportError:
    from yaml import Loader, Dumper

#

def get_custom_generator(env, **kwargs):
    """
    Create a custom populated ParameterGenerator.

    This function recreates the exact same parameter set as the sample LULESH
    specifications. The point of this file is to present an example of how to
    generate custom parameters.

    :returns: A ParameterGenerator populated with parameters.
    """
    p_gen = ParameterGenerator()
    yml = yaml.load(open(sys.argv[-1]).read(), Loader=Loader)
    print(yml)
    p = {}
    labels = {}
    for k, val in yml["global.parameters"].items():
        print(k, val)
        if isinstance(val["values"], (list,tuple)):
            p[k] = set(val["values"])
        else:
            p[k] = [val["values"],]
        labels[k] = val["label"]

    grid = ParameterGrid(p)
    p = {}
    for g in grid:
        for k in g:
            if k not in p:
                p[k] = [g[k],]
            else:
                p[k].append(g[k])
    print("OUT:", p)
    for k, val in p.items():
        p_gen.add_parameter(k, val, labels[k])
    return p_gen
```

**Comment by doutriaux1** [comment]
I take it back it does picks it up! Thanks! Let's close this issue and I will open another ticket about generalizing this.
